src/plugins/core/header_grouping.ts

```diff
@@ -0,0 +1,523 @@
+      return this.doGroupOverlap(group, start, end);
+    });
+  }
+
+  /**
+   * Get all the groups of a sheet in a dimension, and return an array of layers of those groups.
+   *
+   * The layering rules are:
+   * 1) A group containing another group should be on a layer above the group it contains
+   * 2) The widest/highest groups should be on the left/top layer compared to the groups it contains
+   * 3) The group should be on the left/top-most layer possible, barring intersections with other groups (see rules 1 and 2)
+   */
+  getGroupsLayers(sheetId: UID, dimension: Dimension): HeaderGroup[][] {
+    const groups = this.getHeaderGroups(sheetId, dimension);
+    return this.bricksFallingAlgorithm(groups, 0, 0);
+  }
+
+  /**
+   * Get all the groups of a sheet in a dimension, and return an array of layers of those groups,
+   * excluding the groups that are totally hidden.
+   */
+  getVisibleGroupLayers(sheetId: UID, dimension: Dimension): HeaderGroup[][] {
+    const layers: HeaderGroup[][] = this.getGroupsLayers(sheetId, dimension);
+
+    for (const layer of layers) {
+      for (let k = layer.length - 1; k >= 0; k--) {
+        const group = layer[k];
+        if (group.start === 0) {
+          continue;
+        }
+        const headersInGroup = range(group.start - 1, group.end + 1);
+        if (headersInGroup.every((i) => this.getters.isHeaderHiddenByUser(sheetId, dimension, i))) {
+          layer.splice(k, 1);
+        }
+      }
+    }
+
+    return layers.filter((layer) => layer.length > 0);
+  }
+
+  isGroupFolded(sheetId: UID, dimension: Dimension, start: number, end: number): boolean {
+    return this.getHeaderGroup(sheetId, dimension, start, end)?.isFolded || false;
+  }
+
+  isRowFolded(sheetId: UID, row: HeaderIndex) {
+    const groups = this.getters.getHeaderGroups(sheetId, "ROW");
+    for (const group of groups) {
+      if (group.start <= row && row <= group.end) {
+        if (group.isFolded) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  isColFolded(sheetId: UID, col: HeaderIndex) {
+    const groups = this.getters.getHeaderGroups(sheetId, "COL");
+    // might become a performance issue if there are a lot of groups (this is called by isRowHidden).
+    for (const group of groups) {
+      if (group.start <= col && col <= group.end) {
+        if (group.isFolded) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private getGroupId(group: HeaderGroup) {
+    return `${group.start}-${group.end}}`;
+  }
+
+  /**
+   * To get layers of groups, and to add/remove headers from groups, we can see each header of a group as a brick. Each
+   * brick falls down in the pile corresponding to its header, until it hits another brick, or the ground.
+   *
+   * With this abstraction, we can very simply group/ungroup headers from groups, and get the layers of groups.
+   * - grouping headers is done by adding a brick to each header pile
```

Comment:
maybe it'd benefit from some ascii art representation for a better understanding. I tried to ask ChatGPT to make it by explaining the context and giving it the algo, but it failed miserably to output something sensible :laughing: 
I can give it a try if you don't want to.

Reply:
Done 👍